Sources/SourceKitLSP/Swift/SemanticTokens.swift

```diff
@@ -0,0 +1,352 @@
+
+    /// All available modifiers, in order
+    public static let allCases: [Self] = [
+      .declaration,
+      .definition,
+      .readonly,
+      .static,
+      .deprecated,
+      .abstract,
+      .async,
+      .modification,
+      .documentation,
+      .defaultLibrary,
+    ]
+
+    public let rawValue: UInt32
+
+    /// The name of the modifier used by LSP.
+    public var lspName: String? {
+      switch self {
+      case .declaration: return "declaration"
+      case .definition: return "definition"
+      case .static: return "static"
+      case .deprecated: return "deprecated"
+      case .abstract: return "abstract"
+      case .async: return "async"
+      case .modification: return "modification"
+      case .documentation: return "documentation"
+      case .defaultLibrary: return "defaultLibrary"
+      default: return nil
+      }
+    }
+
+    public init(rawValue: UInt32) {
+      self.rawValue = rawValue
+    }
+  }
+}
+
+/// Encodes to the LSP representation of semantic tokens.
+public func encodeToIntArray(semanticTokens tokens: [SemanticToken]) -> [UInt32] {
+  var current = Position(line: 0, utf16index: 0)
+  var rawTokens: [UInt32] = []
+  rawTokens.reserveCapacity(tokens.count * 5)
+
+  for token in tokens {
+    let previous = Position(
+      line: current.line,
+      utf16index: current.line == token.start.line ? current.utf16index : 0
+    )
+    current = token.start
+    rawTokens += [
+      UInt32(token.start.line - previous.line),
+      UInt32(token.start.utf16index - previous.utf16index),
+      UInt32(token.length),
+      token.kind.rawValue,
+      token.modifiers.rawValue
+    ]
+  }
+
+  return rawTokens
+}
+
+/// Decodes the LSP representation of semantic tokens
+public func decodeFromIntArray(rawSemanticTokens rawTokens: [UInt32]) -> [SemanticToken] {
+  var current = Position(line: 0, utf16index: 0)
+  var tokens: [SemanticToken] = []
+  tokens.reserveCapacity(rawTokens.count / 5)
+
+  for i in stride(from: 0, to: rawTokens.count, by: 5) {
+    let lineDelta = Int(rawTokens[i])
+    let charDelta = Int(rawTokens[i + 1])
+    let length = Int(rawTokens[i + 2])
+    let rawKind = rawTokens[i + 3]
+    let rawModifiers = rawTokens[i + 4]
+
+    guard let kind = SemanticToken.Kind(rawValue: rawKind) else { continue }
+    let modifiers = SemanticToken.Modifiers(rawValue: rawModifiers)
+
+    current.line += lineDelta
+
+    if lineDelta == 0 {
+      current.utf16index += charDelta
+    } else {
+      current.utf16index = charDelta
+    }
+
+    tokens.append(SemanticToken(
+      start: current,
+      length: length,
+      kind: kind,
+      modifiers: modifiers
+    ))
+  }
+
+  return tokens
+}
+
+/// Merges the given token arrays into a new token array,
+/// preferring the second array's tokens if duplicate ranges are
+/// found.
+public func mergeSemanticTokens(_ lhs: [SemanticToken], _ rhs: [SemanticToken]) -> [SemanticToken] {
```

Comment:
Yes, overlapping ranges would definitely be preferable, however also less efficient to compute since we can't just use a set there. I think there are specialized data structures like interval trees that let us implement that in linear time, but I'm not sure if it's worth the complexity there.

To me it feels cleaner to keep this method separate from its use in `DocumentTokens`, since it is not really specific to use in a document, even though that is currently the only place where it's used.

Reply:
I see, the performance consideration makes sense. I’m fine with keeping it as-is then.

---

Fair enough. What do you think about making it a (static) member of `SemanticToken`? I don’t like global functions too much because they are less discoverable IMO.

Reply:
It's currently a member of `Array<SemanticToken>` as per your second suggestion, not sure, should it be a static member in `SemanticToken` instead?

Reply:
Oh, I missed that you moved it. Ignore my second comment.